### ISSUE TITLE
runNoReply in java does not work due to toCompletableFuture

### ISSUE THREAD

**Issue by maarest** [rethinkdb/rethinkdb#5938]
Hi,

I don't have a fix for the issue. If you try out this code, you'll see that it will hang:

```
   Connection conn = r.connection().hostname("localhost").port(28015).db("test").connect();
    System.out.println("Connected!");

    r.table("authors").insert(r.hashMap("name", "William Adama")
                    .with("tv_show", "Battlestar Galactica")
                    .with("posts", r.array(
                            r.hashMap("title", "Decommissioning speech")
                                    .with("content", "The Cylon War is long over..."),
                            r.hashMap("title", "We are at war")
                                    .with("content", "Moments ago, this ship received..."),
                            r.hashMap("title", "The new Earth")
                                    .with("content", "The discoveries of the past few days...")
                            )
                    )).run(conn);

    System.out.println("Inserted in Run!");
    r.table("authors").insert(
            r.hashMap("name", "William Adama")
                    .with("tv_show", "Battlestar Galactica")
                    .with("posts", r.array(
                            r.hashMap("title", "Decommissioning speech")
                                    .with("content", "The Cylon War is long over..."),
                            r.hashMap("title", "We are at war")
                                    .with("content", "Moments ago, this ship received..."),
                            r.hashMap("title", "The new Earth")
                                    .with("content", "The discoveries of the past few days...")
                            )
                    )).runNoReply(conn);
    System.out.println("Inserted in RunNoReply!");
```

This is apparently due to this line:
https://github.com/rethinkdb/rethinkdb/blob/next/drivers/java/src/main/java/com/rethinkdb/net/Connection.java#L211
`toCompletableFuture` that never completes in the future

This is I would say something quite important as we really need to have the massive performance improvement brought by runNoReply

Regards,
Stephane

**Comment by bchavez** [comment]
:+1: 

Comparing to the C# driver, seems like the `awaitResponse` check is missing in the Java driver:

https://github.com/bchavez/RethinkDb.Driver/blob/master/Source/RethinkDb.Driver/Net/SocketWrapper.cs#L288

```
SendQuery:
   if( queryIsExpectingResponse ){
      //create the awaiter and keep track of it
   }
   //write the query to the network
```

**Comment by danielmewes** [comment]
Looking into this. Hope we can release a fix in the next days together with the fix for https://github.com/rethinkdb/rethinkdb/issues/5980 .

**Comment by bchavez** [comment]
Cool. I started making some changes but didn't finish. I got about 1/3 there syncing up the C# and Java driver but ran out of time. It didn't seem too hard, just keeping track of the entry point of all the queries and on `Connection` and passing `awaitResponse:true|false` for those `run` methods on `Connection` that expect/not-expect a server response to add (or not add) an awaiter. Should be much easier in Java since the Java driver doesn't have any extra CLR type `Run*` methods like the C# driver.

**Comment by bchavez** [comment]
Here's the `sendQuery` unfinished implementation I had locally:
```
private Future<Response> sendQuery(Query query, Optional<Long> deadline, boolean awaitResponse) {
    // check if response pump is running
    if (!exec.isShutdown() && !exec.isTerminated()) {

        final CompletableFuture<Response> awaiter;
        if(awaitResponse) {
            awaiter = new CompletableFuture<>();
            awaiters.put(query.token, awaiter);
        }
        try {
            lock.lock();
            socket.orElseThrow(() -> new ReqlDriverError("No socket available."))
                    .write(query.serialize());
            if( awaiter != null)
                return awaiter.toCompletableFuture();
            else
                return //return something statically completed in java
                       //CompletableFuture<Response>.Default or something
                       //not sure how to do that yet in java.
        } finally {
            lock.unlock();
        }
    }

    // shouldn't be here
    throw new ReqlDriverError("Can't write query because response pump is not running.");
}
```

**Comment by danielmewes** [comment]
Thanks @bchavez for sharing your code. I ended up implementing this a bit differently via https://github.com/rethinkdb/rethinkdb/commit/91511142dd55328cfae120fc7500ba3e0e1b402f .

This is now in internal code review 3705 by @deontologician .

**Comment by danielmewes** [comment]
Fixed in `next` ca9ca6f and `v2.3.x` 683a031. We will release a new driver soon.

**Comment by danielmewes** [comment]
We just released a new Java driver version 2.3.3 with this fix.

**Comment by maarest** [comment]
Thanks a lot for the quick fix!! Is it uploaded to maven or somewhere else? I can't see it yet
